src/build/build_step.go

```diff
@@ -464,7 +465,20 @@ func calculateAndCheckRuleHash(state *core.BuildState, target *core.BuildTarget)
 	// Set appropriate permissions on outputs
 	if target.IsBinary {
 		for _, output := range target.FullOutputs() {
-			if err := os.Chmod(output, target.OutMode()); err != nil {
+			// Walk through the output,
+			// if the output is a directory,apply output mode to the file instead of the directory
+			err := filepath.Walk(output, func(path string, info os.FileInfo, err error) error {
```

Comment:
please use `fs.Walk` instead of `filepath.Walk` - it is faster (the Go team know the current implementation is suboptimal but can't change the interface on their package)